### PR TITLE
changes the way `gc` detect that an object needs to be deleted

### PR DESCRIPTION
**Pull request by p0lyn0mial**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
This PR resolves the following issue:
1. Create an arbitrary `CRD`
2. Create an arbitrary `CR` of that `CRD`
3. Set a custom `finalizer`
4. Do a [foregreound](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#setting-the-cascading-deletion-policy) delete
5. Wait until the `foregreoundDelete` finalizer disappears
6. Do another [foregreound](https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#setting-the-cascading-deletion-policy)  delete
7. The finalizer(`foregroundDeletion`) will be back

After step `7`, `GC` will not remove the finalizer again and the `CR` is stuck.
The reason why `GC` doesn’t remove/process the second `foregoundDeletion` is https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/garbagecollector/graph_builder.go#L446.
It cannot detect the transition of the object from "not delete" to "delete". Note that the issue applies to any resource, not only CRs.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
do we need an e2e test for that ?

File: pkg/controller/garbagecollector/graph_builder.go [hunk 1]
```diff
@@ -471,13 +468,13 @@ func hasFinalizer(accessor metav1.Object, matchingFinalizer string) bool {
 // this function takes newAccessor directly because the caller already
 // instantiates an accessor for the newObj.
 func startsWaitingForDependentsDeleted(oldObj interface{}, newAccessor metav1.Object) bool {
-	return deletionStarts(oldObj, newAccessor) && hasDeleteDependentsFinalizer(newAccessor)
+	return deletionStarts(oldObj, newAccessor, func(a metav1.Object) bool { return hasDeleteDependentsFinalizer(a) }) && hasDeleteDependentsFinalizer(newAccessor)
```
Comment (liggitt) [inline pkg/controller/garbagecollector/graph_builder.go]: This is pretty confusing to follow... would renaming to deletionStartsWithFinalizer(oldObj, newAccessor, finalizer) and moving all the logic into that method be easier to follow?

```
// if the new object isn't being deleted, or doesn't have the finalizer we're interested in, return false
if !beingDeleted(newAccessor) || !hasFinalizer(newAccessor, finalizer) {
  return false
}

// if the old object is nil, or wasn't being deleted, or didn't have the finalizer, return true
if oldObj == nil {
  return true
}
oldAccessor, err := meta.Accessor(oldObj)
if err != nil {
  utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
  return false
}
return !beingDeleted(oldAccessor) || !hasFinalizer(oldAccessor, finalizer)
```

Comment (p0lyn0mial) [reply pkg/controller/garbagecollector/graph_builder.go]: it looks much better, thanks Jordan :)

**Comment by p0lyn0mial** [comment]
@liggitt I applied your comment and added an integration test, PTAL.

**Pull request merged**